### ISSUE TITLE
Automatically adjust bloom filter size for an sstable

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylla#2024]
If an sstable has very small partitions (perhaps it is just a bunch of tombstones), then the bloom filter may be too large to hold in memory; so the false positive rate should be increased (more false positives) to reduce the impact on memory.

Otherwise, if an sstable has very large partitions, then its bloom filter can be very small.  We can increase the accuracy of the bloom filter since it won't take much memory.

(we can aim for bloom_filter_size = 0.0002 * data_size, or something similar)

**Comment by belliottsmith** [comment]
See also https://issues.apache.org/jira/browse/CASSANDRA-6633 for some related thoughts

**Comment by kostja** [comment]
One possible implementation is to build several bloom filter bitmaps during compaction, with exponential size difference (e.g. x, 2x, 4x, where X is the average size for this sstable), and then choose to store the one which has the optimal false positive ratio. 
The FPR factor can be evaluated as the bloom filter is built.

**Comment by raphaelsc** [comment]
> One possible implementation is to build several bloom filter bitmaps during compaction, with exponential size difference (e.g. x, 2x, 4x, where X is the average size for this table), and then choose to store the one which has the optimal false positive ratio.
> The FPR factor can be evaluated as the bloom filter is built.

This would also need to take into account expiration, as compaction output can be much smaller than input. That's a problem I intend to solve eventually, as compaction involving lots of expired data is generating bloom filters much bigger than needed.

**Comment by mykaul** [comment]
@bhalevy - this is affecting multiple customers. Is there anyone who can work on this?

**Comment by bhalevy** [comment]
Yes, it's just to track the project in our queue.
It's lower priority than the existing projects, though I'd like start advancing it's design ahead of the implementation so we can do some planning sooner rather than later.

I think we should tackle 2 scenarios, initial over-estimate and under-estimate.
That said, I doubt if the in-memory bloom filter is a good choice for dynamic adjustment.
It's relatively easy to "resample" it for downsizing in case it was over-estimated and can be shrinked.
But if it's under-estimated and needs to grow to achieve better resolution, we already lost the original information.

Maybe we can track the tokens dynamically in-memory using a different data structure, like a radix tree,
and generate the bloom filter out of it only when finalizing and sealing the SSTable.

**Comment by nyh** [comment]
> I think we should tackle 2 scenarios, initial over-estimate and under-estimate. That said, I doubt if the in-memory bloom filter is a good choice for dynamic adjustment. It's relatively easy to "resample" it for downsizing in case it was over-estimated and can be shrinked. But if it's under-estimated and needs to grow to achieve better resolution, we already lost the original information.

Note that you didn't lose the information, it's still on in the "index" file - you don't even need to reread the entire data file! 

> Maybe we can track the tokens dynamically in-memory using a different data structure, like a radix tree, and generate the bloom filter out of it only when finalizing and sealing the SSTable.

I think if the problem of oversized or undersized bloom filter, it's not a big issue that flushing an sstable will involve - in those rare cases - yet another step of re-reading the index file and rebuilding the bloom filter. We don't need to waste a lot of memory during the flush just to handle this rare case. Not to mention that I'm not even sure that there won't be cases that we can't even afford this memory use (it's basically analogous to holding the entire index file in memory during the flush).

**Comment by denesb** [comment]
> I think if the problem of oversized or undersized bloom filter, it's not a big issue that flushing an sstable will involve - in those rare cases - yet another step of re-reading the index file and rebuilding the bloom filter. We don't need to waste a lot of memory during the flush just to handle this rare case. Not to mention that I'm not even sure that there won't be cases that we can't even afford this memory use (it's basically analogous to holding the entire index file in memory during the flush).

I agree. I think the best approach is to continue writing bloom filters based on partition count estimates. At sstable seal-time, we check how accurate our estimate was. If it is off by more than some configured margin, we take action. Downsampling is simple. Upsampling can be achieved by re-reading just the index file and re-generating the filter based on it.
The nice thing about this approach is that its optimistic. If things work well, we introduce no overhead, compared to what we do today. We only start to pay at the point where we detect that something is wrong. At this point, the invested effort will help reduce read amplification, so we know it is worth it.

**Comment by bhalevy** [comment]
@mykaul I suggest to reduce the priority of this issue to P2 as it is not blocking

**Comment by mykaul** [comment]
> @mykaul I suggest to reduce the priority of this issue to P2 as it is not blocking

We have several customers impacted by it. It's not a blocker and can be pushed to 6.1, but certainly looks like something we need to prioritize.

**Comment by bhalevy** [comment]
@denesb, @lkshminarayanan what's left to be done here?

**Comment by lkshminarayanan** [comment]
> @denesb, @lkshminarayanan what's left to be done here?

There is this PR that folds down bloom filters if possible - https://github.com/scylladb/scylladb/pull/18849

But after https://github.com/scylladb/scylladb/pull/19190, I'm not sure if we still need it.

**Comment by denesb** [comment]
We also didn't do size-based reduction. We currently regenerate the filter if it was based on a wrong partition estimate. There is also the case where the filter was based on the correct estimate, but the sstable has a lot of partitions and the filter is too large.

**Comment by bhalevy** [comment]
Demoting to P2 since it affects only certain workloads (small partitions, lots of deletions)

**Comment by dani-tweig** [comment]
This one has been traveling with us for a long while.
Do we need it?

**Comment by denesb** [comment]
> This one has been traveling with us for a long while. Do we need it? Will we find the time to develop it?

@lkshminarayanan is scheduled to start working on this in Q2.

**Comment by lkshminarayanan** [comment]
> > This one has been traveling with us for a long while. Do we need it? Will we find the time to develop it?
> 
> [@lkshminarayanan](https://github.com/lkshminarayanan) is scheduled to start working on this in Q2.

There is one more bloom filter specific issue remaining before we move on to https://github.com/scylladb/scylladb/issues/21897 :

If a bloom filter is too large for memory even after being built with the correct partition count, we should reduce its size. This also means that the effective bloom filter false positive rate will increase for that particular sstable.

Before sealing the sstable, we already check if the Bloom filter is built with a good partition estimate. I plan to update this to also check whether the final Bloom filter will be too large for memory, and if so, rebuild it with a reduced size.

@denesb @avikivity With the above context, what is a good `disk:bloom_filter_size` ratio limit that we should check for?  If a bloom filter grows beyond this limit, it will be reduced to maintain the ratio.

**Comment by avikivity** [comment]
We don't want to do that for the last bloom filter we create, rather we want to do it to all resident filters.

As to a ratio, we have a config item for it already, no?

**Comment by denesb** [comment]
I remember that we were aiming for a sstable:data_size/sstable:bloom_filter_size ratio, this is something we can apply to individual sstables, without having to look at all filters.

One thing to keep in mind: after reducing a bloom filter by a certain factor, the filter becomes all 1s and is essentially useless (it will return [false] positive for all queries). At this point we can just drop the filter because it stops working.

**Comment by avikivity** [comment]
> I remember that we were aiming for a sstable:data_size/sstable:bloom_filter_size ratio, this is something we can apply to individual sstables, without having to look at all filters.

It may be wasteful. If we have an sstable with lots of tombstones (say a materialized view sstable), it can be balanced with with a table that has relatively larger partitions.

Although: typically materialized view tombstones are row tombstones, not partition tombstones.

> 
> One thing to keep in mind: after reducing a bloom filter by a certain factor, the filter becomes all 1s and is essentially useless (it will return [false] positive for all queries). At this point we can just drop the filter because it stops working.

Yes.

**Comment by mykaul** [comment]
@denesb , @lkshminarayanan - is this still on track for 2025.4?

**Comment by lkshminarayanan** [comment]
> [@denesb](https://github.com/denesb) , [@lkshminarayanan](https://github.com/lkshminarayanan) - is this still on track for 2025.4?

Hi! No, we decided to collect some statistics on bloom filter sizes by simulating large production sized clusters. Only after that we can decide how to proceed with this.

**Comment by mykaul** [comment]
> > [@denesb](https://github.com/denesb) , [@lkshminarayanan](https://github.com/lkshminarayanan) - is this still on track for 2025.4?
> 
> Hi! No, we decided to collect some statistics on bloom filter sizes by simulating large production sized clusters. Only after that we can decide how to proceed with this.

We have Scylla cloud, we can gather statistics from there. Where can I track the progress of this item?

I regretfully moved the item to 2026.1. Let's ensure we make progress please.

**Comment by lkshminarayanan** [comment]
There is a document at https://scylladb.atlassian.net/wiki/spaces/RND/pages/95257057/Bloom+Filter+Resizing+Requirements+Document , though the goal is no longer just resizing. Once we collect the stats, we’ll need to decide how to proceed.

I've already mailed the cloud team and received the candidate clusters, which needs to be reviewed to collect the stats.
I'll get to it soon.

**Comment by denesb** [comment]
It was decided that we will not pursue automatic bloom filter resizing. Closing this issue as wontfix.
